traversal/executor/constraint_iterator.rs

```diff
@@ -0,0 +1,277 @@
+        match self.iterate_mode {
+            SortedIterateMode::UnboundSortedFrom => {
+                let first_from_type = self.owner_attribute_types.first_key_value().unwrap().0;
+                let last_key_from_type = self.owner_attribute_types.last_key_value().unwrap().0;
+                let key_range = KeyRange::new_inclusive(first_from_type.as_object_type(), last_key_from_type.as_object_type());
+                let filter_fn = self.filter_fn.has_filter();
+                // TODO: we could cache the range byte arrays computed inside the thing_manager, for this case
+                let iterator: Filter<HasIterator, Arc<HasFilterFn>> = thing_manager
+                    .get_has_from_type_range_unordered(snapshot, key_range)
+                    .filter::<_, HasFilterFn>(filter_fn);
+                Ok(ConstraintIterator::HasUnboundedSortedFrom(iterator))
+            }
+            SortedIterateMode::UnboundSortedTo => {
+                // debug_assert!(self.owner_cache.is_some());
+                // let mut iterators = Vec::new();
+                // for iter in self.owner_cache.as_ref().unwrap().iter().map(|object|
+                //     object.get_has_types_range_unordered(snapshot, thing_manager, self.attribute_types.iter().cloned())
+                // ) {
+                //     let iter: HasAttributeIterator = iter?;
+                //
+                // }
+                //
+                // let merged = kmerge_by(iterators, |(attr_1, _), (attr_2, _)| attr_1 < attr_2);
+                // let unique = merged.dedup_by(|(attr_1, _)| attr_1);
+                // let single_counts = unique.map(|(attr, count)| (attr, 1));
+                //
+                // // ConstraintIterator::HasUnboundSortedTo(single_counts)
+                todo!()
+            }
+            SortedIterateMode::BoundFromSortedTo => {
+                todo!()
+            }
+        }
+    }
+}
+
+struct HasReverseProvider {
+    has: Has<Position>,
+    iterate_mode: SortedIterateMode,
+}
+
+impl HasReverseProvider {
+    pub(crate) fn get_iterator(&self, row: &Row) -> ConstraintIterator {
+        todo!()
+    }
+}
+
+struct RolePlayerProvider {
+    role_player: RolePlayer<Position>,
+    iterate_mode: SortedIterateMode,
+}
+
+impl RolePlayerProvider {
+    pub(crate) fn get_iterator(&self, row: &Row) -> ConstraintIterator {
+        todo!()
+    }
+}
+
+struct RolePlayerReverseProvider {
+    role_player: RolePlayer<Position>,
+    iterate_mode: SortedIterateMode,
+}
+
+impl RolePlayerReverseProvider {
+    pub(crate) fn get_iterator(&self, row: &Row) -> ConstraintIterator {
+        todo!()
+    }
+}
+
+struct FunctionCallBindingProvider {
+    function_call_binding: FunctionCallBinding<Position>,
+}
+
+struct ComparisonProvider {
+    comparison: Comparison<Position>,
+}
+
+struct ComparisonReverseProvider {
+    comparision: Comparison<Position>,
+}
+
+pub(crate) enum ConstraintIterator {
```

Comment:
Goal: explicitly encode all the possible iterators we'll need, without doing any heap allocations.... let's see if we can!